### PR TITLE
Template update, big DSL2 modules update

### PR DESCRIPTION
**Pull request by ewels**
My attempt at getting the DSL2 port of this pipeline over the line.

A lot has changed since @phue originally worked on this. For starters, the DSL2 template is now officially released. This meant that I merged in the `TEMPLATE` commits from the sync (which had a lot of merge conflicts as I think it was all copied in before without git history). There are also a lot of changes in modules syntax, and many previously local modules are now merged in to the central repo.

WIP - been manually trying to merge this all but untested so far, so still quite a way to go.

**Note: This PR is to the `dsl2` branch and not `dev`**
I hope that when this is merged we can then get on to merging #199 soon after and then release.

When this is merged, it will encompass #220 and so that will close automatically immediately.

- [x] Fix merge conflicts from template sync
- [x] Remove local modules and replace with nf-core/modules
- [x] Strip out all modules and reinstall with new `modules.json`
- [ ] Check that there aren't any now-dead files left hanging after the template update
- [ ] Go through linting errors
- [ ] Go through test pipeline errors
- [ ] Final test for review

### CODE REVIEW

File: modules/nf-core/modules/bwameth/align/main.nf [hunk 1]
```diff
@@ -19,22 +19,19 @@ process BWAMETH_ALIGN {
...
     script:
-    def software   = getSoftwareName(task.process)
     def prefix     = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
     def read_group = meta.read_group ? "-R ${meta.read_group}" : ""
     """
     INDEX=`find -L ${index} -name "*.bwameth.c2t" | sed 's/.bwameth.c2t//'`
 
-    # Modify the timestamps so that bwameth doesn't complain about building the index
-    touch -c -- *
-
```
Comment (ewels) [inline modules/nf-core/modules/bwameth/align/main.nf:37]: Not sure if we can get away with removing this. I added it after a lot of problems with bwameth.

@phue if you're reading this do you remember why it was removed in the local -> nf-core/modules switch?

**Review by ewels — COMMENTED** [review]
Self-review

**Comment by phue** [comment]
Hi @ewels,

Thanks for picking this up again, I hadn't looked at this for a (quite a) while.

Just as a heads up, there are few caveats regarding the removal of local modules:

There were 2 main reasons why I went with local modules for this:
i) relates to the discussion here: https://github.com/nf-core/modules/pull/239#discussion_r594618090
To keep pipeline-specific code out of `nf-core/modules`, I went with just keeping those memory/cpu settings in the local modules as the logic is not only scary but also depends on library presets as they are defined in the pipeline code.
But it is not optimal, because the generic module lacks multithreading completely because of that.

ii) relates to https://github.com/nf-core/methylseq/issues/181 (some discussion here: https://github.com/nf-core/modules/pull/129#issuecomment-774997513)
The current implementation on the dsl2 branch relies on the `meta` map being attached to not only samples, but also the reference genome to enable mapping against multiple pseudogenomes (or assemblies) in one pipeline run.
See here for example: https://github.com/ewels/nf-core-methylseq/blob/dsl2-template-merge/subworkflows/local/bismark.nf#L54-L61

File: modules/nf-core/modules/bwameth/align/main.nf (hunk 1 above)
Comment (phue) [reply modules/nf-core/modules/bwameth/align/main.nf:37]: I see that thing for the first time :sweat_smile: 
Looks like you added that in June https://github.com/nf-core/methylseq/commit/f6b18b352f087e08cd265644c14486505586e517

But it doesn't seem to be on master

Comment (ewels) [reply modules/nf-core/modules/bwameth/align/main.nf:37]: Right, yeah then that'd do it 😅  Sorry, getting my timeline mixed up..

Was added into `dev` in June, but you're right - not out in a release yet: https://github.com/nf-core/methylseq/pull/217

Comment (ewels) [reply modules/nf-core/modules/bwameth/align/main.nf:37]: Added to the module: https://github.com/nf-core/modules/pull/1037

**Comment by ewels** [comment]
Hi @phue!

Great to hear from you 😀 Massive kudos for the work you did here with the DSL2 conversion, sorry it's taken me so long to get around to helping you with it.

i) Ok so I think I'm starting to understand this now. I get the reasoning for keeping the module as simple and generic as possible. But does the pipeline need to have a local module for this? Can it not just append to the `$options.args` like other flags and use the central module?

ii) Ahhh, you mean this subtle little difference?

```diff
-    tuple val(meta), path(bam), path(index)
+    tuple val(meta), path(bam)
+    path index
```

I totally missed that.. Right, and the objection for having it in nf-core/modules with that style is that it's not consistent with other modules / too complex?

Ok I see both standpoints. I kind of want to avoid having local versions of modules that are also available in the central nf-core/modules repo though. I think that it's inevitable that there will be feature drift between them and it kind of loses the advantages of having the DSL2 modules. I'll take it up on Slack and see if there's a way that we can resolve this.

**Comment by phue** [comment]
> sorry it's taken me so long to get around to helping you with it.
No worries, great that you brought it up again so we can have this discussion and get this thing finalized :+1:

> i) Can it not just append to the `$options.args` like other flags and use the central module?

Yes, that's probably the cleaner option. Having a groovy function to do these memory/core calculations would be even better. I just went with the local modules because they were anyway required for `ii)`
 
> Right, and the objection for having it in nf-core/modules with that style is that it's not consistent with other modules / too complex?

I don't think it's complex or anything, consistency is what mattered in the discussion back then. It would have required to extend the `meta` map concept beyond samples but also to references/indices etc. Doing this change now across modules would be even more cumbersome as it would likely break existing dsl2 pipeline code.

File: modules/nf-core/modules/bwameth/align/main.nf (hunk 1 above)
Comment (ewels) [reply modules/nf-core/modules/bwameth/align/main.nf]: Thanks for the merge! Pipeline updated with new version of the module.. ✅

**Comment by ewels** [comment]
Reminder to myself: Need to implement the new versioning code. Docs: https://nf-co.re/developers/adding_modules#new-module-guidelines-and-pr-review-checklist + check rnaseq pipeline.

**Pull request closed without merging**